File: BACKEND LARAVEL/app/Http/Controllers/PendudukCon.php
<?php

namespace App\Http\Controllers;

use App\Models\Penduduk;
use Illuminate\Http\Request;
use App\Http\Resources\PendudukRes;
use Illuminate\Support\Facades\Validator;
use App\Models\User;
use Illuminate\Support\Facades\DB;
use Illuminate\Pagination\LengthAwarePaginator;

class PendudukCon extends Controller
{
    /**
     * Display a listing of the resource.
     */
    public function getPenduduk(Request $request)
    {
        $tampil=$request->input('item');
        $total= Penduduk::count();
        $role =$request->user()->level;
        $datas;
        if ($role=='enum') {
            $id = $request->user()->id;
            $total= Penduduk::where('user_id',$id)->count();
            $datas = Penduduk::where('user_id',$id)->with('keluarga','user')->paginate($tampil);
        }
        else if($role=='admin'||$role=='superAdmin'){
            $datas = Penduduk::with('keluarga','user')->paginate($tampil);
        }
              $hasils= $datas->getCollection()->reduce(
                function ($items, $data){

                    $agama=penduduk::agama($data->agama);
                    $hub=penduduk::hub_kel($data->hub_kel);
                    $stat_kwn=penduduk::stat_kawin($data->stat_kawin);
                    $kelamin=penduduk::kelamin($data->kelamin);
                    $pendidikan=penduduk::pendidikan($data->pendidikan);
                    $pekerjaan=penduduk::pekerjaan($data->pekerjaan);
                    $stat=penduduk::stat($data->stat);
                    $warga=penduduk::warga_neg($data->warga_neg);


                    $items[] = [
                        'id'=>$data->id,
                        'kels_id'=>$data->kels_id,
                        'nomer_kk'=>$data->keluarga->no_kk,
                        'nik' =>$data->nik,
                        'nama'=>$data->nama,
                        'tmp_lhr'=>$data->tmp_lhr,
                        'tgl_lhr'=>$data->tgl_lhr,
                        'kelamin'=>$kelamin,
                        'stat_kawin'=>$stat_kwn,
                        'hub_kel'=>$hub,
                        'warga_neg'=>$warga,
                        'agama' => $agama,
                        'pendidikan'=>$pendidikan,
                        'pekerjaan'=>$pekerjaan,
                        'ayah'=>$data->ayah,
                        'ibu'=>$data->ibu,
                        'kepala_kel'=>$data->keluarga->kk_nama,
                        'no_hp'=>$data->no_hp,
                        'domisili'=>$data->domisili,
                        'stat' =>$stat,
                        'user_id'=>$data->user->name
                    ];
                    return $items;
                },
            );
            $currentPage = LengthAwarePaginator::resolveCurrentPage();
            $perPage = $tampil;
            $currentItems = array_slice($hasils, ($currentPage - 1) * $perPage, $perPage);
            $paginatedItems = new LengthAwarePaginator($hasils, $total , $perPage);


            // $result = [
            //     'data' => $hasill->items(),
            //     'pagination' => [
            //         'current_page' => $posts->currentPage(),
            //         'per_page' => $posts->perPage(),
            //         'total' => $posts->total(),
            //         'last_page' => $posts->lastPage(),
            //         'next_page_url' => $posts->nextPageUrl(),
            //         'prev_page_url' => $posts->previousPageUrl(),
            //         'first_page_url' => $posts->url(1),
            //         'last_page_url' => $posts->url($posts->lastPage()),
            //     ],
            //     'total_word_count' => $totalWordCount,
            // ];

// return $posts;
        return response()->json([
            'success' => true,
            'data' => $paginatedItems
        ], 200);

    }
    public function getbyID($id){
        $dt= Penduduk::where('id',$id)->get();
        return response()->json([
            'data' => $dt,
        ], 200);
    }
    public function latest(Request $request)
    {
        $role =$request->user()->level;
        $datas;
        if ($role=='enum') {
            $id = $request->user()->id;
            $datas = Penduduk::where('user_id',$id)->latest()->take(5)->get();
        }
        else if($role=='admin'||$role=='superAdmin'){
            $datas = Penduduk::latest()->take(5)->get();
        }
        $hasil = $datas->reduce(
            function ($items, $data){

                $agama=penduduk::agama($data->agama);
                $hub=penduduk::hub_kel($data->hub_kel);
                $stat_kwn=penduduk::stat_kawin($data->stat_kawin);
                $kelamin=penduduk::kelamin($data->kelamin);
                $pendidikan=penduduk::pendidikan($data->pendidikan);
                $pekerjaan=penduduk::pekerjaan($data->pekerjaan);
                $stat=penduduk::stat($data->stat);
                $warga=penduduk::warga_neg($data->warga_neg);


                $items[] = [
                    'id'=>$data->id,
                    'kk'=>$data->keluarga->no_kk.' / '.$data->keluarga->kk_nama,
                    'nik' =>$data->nik,
                    'nama'=>$data->nama,
                    'kepala_kel'=>$data->keluarga->kk_nama,
                    'status' =>$stat,
                    'user_id'=>$data->user->name

                ];
                return $items;
            }, []
        );


    return response()->json($hasil);
    }
    public function showPenduduk($id)
    {
        $dt_produk = Penduduk::where('id',$id)->get();

        $hasil = Penduduk::where('id',$id)->get()->reduce(
            function ($items, $data){

                $agama=penduduk::agama($data->agama);
                $hub=penduduk::hub_kel($data->hub_kel);
                $stat_kwn=penduduk::stat_kawin($data->stat_kawin);
                $kelamin=penduduk::kelamin($data->kelamin);
                $pendidikan=penduduk::pendidikan($data->pendidikan);
                $pekerjaan=penduduk::pekerjaan($data->pekerjaan);
                $stat=penduduk::stat($data->stat);
                $warga=penduduk::warga_neg($data->warga_neg);


                $items[] = [
                    'id'=>$data->id,
                    'nomer_kk'=>$data->keluarga->no_kk,
                    'nik' =>$data->nik,
                    'nama'=>$data->nama,
                    'tmp_lhr'=>$data->tmp_lhr,
                    'tgl_lhr'=>$data->tgl_lhr,
                    'kelamin'=>$kelamin,
                    'stat_kawin'=>$stat_kwn,
                    'hub_kel'=>$hub,
                    'warga_neg'=>$warga,
                    'agama' => $agama,
                    'pendidikan'=>$pendidikan,
                    'pekerjaan'=>$pekerjaan,
                    'ayah'=>$data->ayah,
                    'ibu'=>$data->ibu,
                    'kepala_kel'=>$data->keluarga->kk_nama,
                    'no_hp'=>$data->no_hp,
                    'domisili'=>$data->domisili,
                    'stat' =>$stat,
                    'user_id'=>$data->user->name

                ];
                return $items;
            }, []
        );


    return response()->json([
        'data' => $hasil,
    ], 200);


        return response()->json($dt_produk);
    }
    public function addPenduduk(Request $req){

        $validator= Validator::make($req->all(),[
            'kels_id'=>'required',
            'nik'=>'required',
            'nama'=>'required',
            'tmp_lhr'=>'required',
            'tgl_lhr'=>'required',
            'kelamin'=>'required',
            'stat_kawin'=>'required',
            'hub_kel'=>'required',
            'warga_neg'=>'required',
            'agama'=>'required',
            'pendidikan'=>'required',
            'pekerjaan'=>'required',
            'ayah'=>'required',
            'ibu'=>'required',
            'no_hp'=>'required',
            'domisili'=>'required',
            'stat'=>'required',
            'user_id'=>'required'


        ]);
        if ($validator->fails()) {
            return response()->json(['valid'=>false,'massage'=>'pastikan form sudah terisi dengan benar']);
        }
        $save=Penduduk::create([
            'kels_id'=>$req->get('kels_id'),
            'nik'=>$req->get('nik'),
            'nama'=>$req->get('nama'),
            'tmp_lhr'=>$req->get('tmp_lhr'),
            'tgl_lhr'=>$req->get('tgl_lhr'),
            'kelamin'=>$req->get('kelamin'),
            'stat_kawin'=>$req->get("stat_kawin"),
            'hub_kel'=>$req->get('hub_kel'),
            'warga_neg'=>$req->get('warga_neg'),
            'agama'=>$req->get('agama'),
            'pendidikan'=>$req->get('pendidikan'),
            'pekerjaan'=>$req->get('pekerjaan'),
            'ayah'=>$req->get('ayah'),
            'ibu'=>$req->get('ibu'),
            'no_hp'=>$req->get('no_hp'),
            'domisili'=>$req->get('domisili'),
            'stat'=>$req->get('stat'),
            'user_id'=>$req->get('user_id')


        ]);

        if ($save) {
            return response()->json(['valid'=>true,'massage'=>'sukses menambah Penduduk']);
        }
        else{
            return response()->json(['valid'=>false,'massage'=>'gagal menambah Penduduk']);
        }

    }

    public function updatePenduduk(Request $req, $id){

        $validator=Validator::make($req->all(),[
            'kels_id'=>'required',
            'nik'=>'required',
            'nama'=>'required',
            'tmp_lhr'=>'required',
            'tgl_lhr'=>'required',
            'kelamin'=>'required',
            'stat_kawin'=>'required',
            'hub_kel'=>'required',
            'warga_neg'=>'required',
            'agama'=>'required',
            'pendidikan'=>'required',
            'pekerjaan'=>'required',
            'ayah'=>'required',
            'ibu'=>'required',
            'no_hp'=>'required',
            'domisili'=>'required',
            'stat'=>'required',
            'user_id'=>'required'

        ]);

        if ($validator->fails()) {
            return response()->json($validator->errors()->toJson());
        }
        $save=Penduduk::where('id', $id)->update($req->all());
        if ($save) {
            return response()->json(['status'=>true,'massage'=>'sukses edit penduduk']);
        }
        else{
            return response()->json(['status'=>false,'massage'=>'gagal edit penduduk']);
        }

    }
    public function destroy($id)
    {
        $penduduk = Penduduk::find($id);

        if ($penduduk) {
            $penduduk->delete();
            return response()->json(['message' => 'Data penduduk berhasil dihapus.'], 200);
        } else {
            return response()->json(['message' => 'Data penduduk tidak ditemukan.'], 404);
        }
    }
}

File: backend_golang/controllers/pendudukCon.go
package controllers

import (
	"backend_golang/config"
	"backend_golang/models"
	"backend_golang/setup"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

func GetPenduduk(c *gin.Context) {
	// Mengambil data user yang sedang login
	userID, exists := c.Get("user")
	if !exists {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found", "status": false})
		return
	}

	var user models.User
	if err := setup.DB.First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found", "status": false})
		return
	}

	var penduduk []models.Penduduk
	var err error
	var total int64

	if user.Level == "admin" {
		err = setup.DB.Preload("User").Preload("Keluarga").Order("created_at DESC").Find(&penduduk).Error
	} else {
		err = setup.DB.Preload("User").Preload("Keluarga").Where("user_id = ?", user.Id).Order("created_at DESC").Find(&penduduk).Error
	}

	setup.DB.Model(&models.Penduduk{}).Count(&total)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": false,
			"error":  "Gagal mengambil data: " + err.Error(),
		})
		return
	}

	if len(penduduk) == 0 {
		c.JSON(http.StatusOK, gin.H{
			"status":  true,
			"data":    []map[string]interface{}{},
			"message": "Data kosong",
		})
		return
	}

	formattedPenduduks := make([]gin.H, len(penduduk))
	for i, penduduk := range penduduk {

		formattedPenduduks[i] = gin.H{
			"id":         penduduk.Id,
			"kels_id":    penduduk.KelsId,
			"nomer_kk":   penduduk.Keluarga.NoKk,
			"nik":        penduduk.Nik,
			"nama":       penduduk.Nama,
			"tmp_lhr":    penduduk.TmpLahir,
			"tgl_lhr":    penduduk.TglLahir,
			"kelamin":    config.GetKelamin(int(penduduk.Kelamin)),
			"stat_kawin": config.GetStatusKawin(int(penduduk.StatKawin)),
			"hub_kel":    config.GetHubunganKeluarga(int(penduduk.HubKel)),
			"warga_neg":  config.GetWargaNegara(int(penduduk.WargaNeg)),
			"agama":      config.GetAgama(int(penduduk.Agama)),
			"pendidikan": config.GetPendidikan(int(penduduk.Pendidikan)),
			"pekerjaan":  config.GetPekerjaan(int(penduduk.Pekerjaan)),
			"ayah":       penduduk.Ayah,
			"ibu":        penduduk.Ibu,
			"kepala_kel": penduduk.Keluarga.KkNama,
			"no_hp":      penduduk.NoHp,
			"domisili":   penduduk.Domisili,
			"stat":       config.GetStatus(int(penduduk.Status)),
			"user_id":    penduduk.User.Id,
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"user":  user,
		"data":  formattedPenduduks,
		"total": total,
	})
}

func GetLatestPenduduk(c *gin.Context) {
	// Mengambil data user yang sedang login
	userID, exists := c.Get("user")
	if !exists {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found", "status": false})
		return
	}

	var user models.User
	if err := setup.DB.First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found", "status": false})
		return
	}

	var penduduk []models.Penduduk
	var err error

	if user.Level == "admin" {
		err = setup.DB.Preload("User").Preload("Keluarga").Order("created_at DESC").Find(&penduduk).Error
	} else {
		err = setup.DB.Preload("User").Preload("Keluarga").Where("user_id = ?", user.Id).Order("created_at DESC").Find(&penduduk).Error
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": false,
			"error":  "Gagal mengambil data: " + err.Error(),
		})
		return
	}

	if len(penduduk) == 0 {
		c.JSON(http.StatusOK, gin.H{
			"status":  true,
			"data":    []map[string]interface{}{},
			"message": "Data kosong",
		})
		return
	}

	formattedPenduduks := make([]gin.H, len(penduduk))
	for i, penduduk := range penduduk {

		formattedPenduduks[i] = gin.H{
			"id":         penduduk.Id,
			"nomer_kk":   penduduk.Keluarga.NoKk,
			"nik":        penduduk.Nik,
			"nama":       penduduk.Nama,
			"kepala_kel": penduduk.Keluarga.KkNama,
			"stat":       config.GetStatus(int(penduduk.Status)),
			"user_id":    penduduk.User.Name,
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"user":  user,
		"data":  formattedPenduduks,
	})
}

func GetPendudukByID(c *gin.Context) {
	id := c.Param("id")

	var penduduk models.Penduduk
	if err := setup.DB.Preload("Keluarga").Preload("User").First(&penduduk, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Data tidak ditemukan"})
		return
	}

	hasil := map[string]interface{}{
		"id":         penduduk.Id,
		"nomer_kk":   penduduk.Keluarga.NoKk,
		"nik":        penduduk.Nik,
		"nama":       penduduk.Nama,
		"tmp_lhr":    penduduk.TmpLahir,
		"tgl_lhr":    penduduk.TglLahir,
		"kelamin":    config.GetKelamin(int(penduduk.Kelamin)),
		"stat_kawin": config.GetStatusKawin(int(penduduk.StatKawin)),
		"hub_kel":    config.GetHubunganKeluarga(int(penduduk.HubKel)),
		"warga_neg":  config.GetWargaNegara(int(penduduk.WargaNeg)),
		"agama":      config.GetAgama(int(penduduk.Agama)),
		"pendidikan": config.GetPendidikan(int(penduduk.Pendidikan)),
		"pekerjaan":  config.GetPekerjaan(int(penduduk.Pekerjaan)),
		"ayah":       penduduk.Ayah,
		"ibu":        penduduk.Ibu,
		"kepala_kel": penduduk.Keluarga.KkNama,
		"no_hp":      penduduk.NoHp,
		"domisili":   penduduk.Domisili,
		"stat":       config.GetStatus(int(penduduk.Status)),
		"user_id":    penduduk.User.Name,
	}

	c.JSON(http.StatusOK, gin.H{"data": hasil})
}

func AddPenduduk(c *gin.Context) {
	var input struct {
		KelsId     int64     `json:"kels_id" binding:"required"`
		Nik        int64     `json:"nik" binding:"required"`
		Nama       string    `json:"nama" binding:"required"`
		TmpLahir   string    `json:"tmp_lhr" binding:"required"`
		TglLahir   time.Time `json:"tgl_lhr" binding:"required"`
		Kelamin    int8      `json:"kelamin" binding:"required"`
		StatKawin  int8      `json:"stat_kawin" binding:"required"`
		HubKel     int8      `json:"hub_kel" binding:"required"`
		WargaNeg   int8      `json:"warga_neg" binding:"required"`
		Agama      int8      `json:"agama" binding:"required"`
		Pendidikan int8      `json:"pendidikan" binding:"required"`
		Pekerjaan  int8      `json:"pekerjaan" binding:"required"`
		Ayah       string    `json:"ayah" binding:"required"`
		Ibu        string    `json:"ibu" binding:"required"`
		NoHp       string    `json:"no_hp" binding:"required"`
		Domisili   int8      `json:"domisili" binding:"required"`
		Status     int8      `json:"stat" binding:"required"`
		UserId     int64     `json:"user_id" binding:"required"`
	}

	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"valid":   false,
			"message": "Pastikan form sudah terisi dengan benar",
		})
		return
	}

	penduduk := models.Penduduk{
		KelsId:     input.KelsId,
		Nik:        input.Nik,
		Nama:       input.Nama,
		TmpLahir:   input.TmpLahir,
		TglLahir:   input.TglLahir,
		Kelamin:    input.Kelamin,
		StatKawin:  input.StatKawin,
		HubKel:     input.HubKel,
		WargaNeg:   input.WargaNeg,
		Agama:      input.Agama,
		Pendidikan: input.Pendidikan,
		Pekerjaan:  input.Pekerjaan,
		Ayah:       input.Ayah,
		Ibu:        input.Ibu,
		NoHp:       input.NoHp,
		Domisili:   input.Domisili,
		Status:     input.Status,
		UserId:     input.UserId,
	}

	if err := setup.DB.Create(&penduduk).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"valid":   false,
			"message": "Gagal menambah penduduk",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"valid":   true,
		"message": "Sukses menambah penduduk",
	})
}

func UpdatePenduduk(c *gin.Context) {
	id := c.Param("id")

	var penduduk models.Penduduk
	if err := setup.DB.First(&penduduk, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"status": false,
			"message": "Data tidak ditemukan",
		})
		return
	}

	// Update field satu per satu
	if err := c.ShouldBindJSON(&penduduk); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": false,
			"message": "Pastikan form sudah terisi dengan benar",
		})
		return
	}

	// Gunakan Save() untuk memperbarui semua field
	if err := setup.DB.Save(&penduduk).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": false,
			"message": "Gagal mengupdate penduduk",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"status": true,
		"message": "Sukses mengupdate penduduk",
	})
}

func DeletePenduduk(c *gin.Context) {
	id := c.Param("id")

	var penduduk models.Penduduk
	if err := setup.DB.First(&penduduk, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"message": "Data penduduk tidak ditemukan",
		})
		return
	}

	if err := setup.DB.Delete(&penduduk).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"message": "Gagal menghapus data penduduk",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Data penduduk berhasil dihapus",
	})
}
